File: augur-engine/src/publish/git-push.ts
import { simpleGit } from 'simple-git';

/** Commit and push new content to the augur site repo/branch */
export async function commitAndPush(
  siteDir: string,
  message: string,
  branch?: string,
): Promise<void> {
  const targetBranch = branch ?? process.env['SITE_BRANCH'] ?? 'augur_news';
  const git = simpleGit(siteDir);

  // Ensure we're on the right branch
  const currentBranch = (await git.branchLocal()).current;
  if (currentBranch !== targetBranch) {
    try {
      await git.checkout(targetBranch);
    } catch {
      await git.checkoutLocalBranch(targetBranch);
    }
  }

  // Stage all changes
  await git.add('.');

  // Check if there are actual changes
  const status = await git.status();
  if (status.files.length === 0) {
    console.log('[git] no changes to commit');
    return;
  }

  // Commit
  await git.commit(message);
  console.log(`[git] committed: ${message} (${status.files.length} files)`);

  // Push with retry
  const maxRetries = 4;
  const backoff = [2000, 4000, 8000, 16000];

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      await git.push('origin', targetBranch, ['--set-upstream']);
      console.log(`[git] pushed to origin/${targetBranch}`);
      return;
    } catch (err) {
      if (attempt < maxRetries) {
        const delay = backoff[attempt];
        console.warn(`[git] push failed (attempt ${attempt + 1}/${maxRetries + 1}), retrying in ${delay}ms...`);
        await new Promise((r) => setTimeout(r, delay));
      } else {
        throw new Error(`[git] push failed after ${maxRetries + 1} attempts: ${err}`);
      }
    }
  }
}

File: augur-engine/src/publish/git_push.py
"""Git commit and push for the augur site repo."""

from __future__ import annotations

import asyncio
import logging
import os

log = logging.getLogger("augur.publish")


async def commit_and_push(
    site_dir: str,
    message: str,
    branch: str | None = None,
) -> None:
    """Commit and push new content to the augur site repo/branch."""
    target_branch = branch or os.environ.get("SITE_BRANCH", "augur_news")

    async def _run(cmd: list[str]) -> tuple[int, str, str]:
        proc = await asyncio.create_subprocess_exec(
            *cmd,
            cwd=site_dir,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await proc.communicate()
        return proc.returncode or 0, stdout.decode(), stderr.decode()

    # Ensure we're on the right branch
    rc, current, _ = await _run(["git", "rev-parse", "--abbrev-ref", "HEAD"])
    current = current.strip()
    if current != target_branch:
        rc, _, _ = await _run(["git", "checkout", target_branch])
        if rc != 0:
            await _run(["git", "checkout", "-b", target_branch])

    # Stage all changes
    await _run(["git", "add", "."])

    # Check if there are actual changes
    rc, status_out, _ = await _run(["git", "status", "--porcelain"])
    if not status_out.strip():
        log.info("no changes to commit")
        return

    # Commit
    await _run(["git", "commit", "-m", message])
    log.info("committed: %s", message)

    # Push with retry (exponential backoff)
    backoff = [2, 4, 8, 16]
    for attempt in range(5):
        rc, _, stderr = await _run(
            ["git", "push", "-u", "origin", target_branch]
        )
        if rc == 0:
            log.info("pushed to origin/%s", target_branch)
            return
        if attempt < 4:
            delay = backoff[attempt]
            log.warning(
                "push failed (attempt %d/5), retrying in %ds...", attempt + 1, delay
            )
            await asyncio.sleep(delay)

    raise RuntimeError(f"push failed after 5 attempts: {stderr}")
